File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/sorting/merge_sort.py
def merge_sort(values: list[int]) -> list[int]:
    if len(values) <= 1:
        return values[:]

    midpoint = len(values) // 2
    left = merge_sort(values[:midpoint])
    right = merge_sort(values[midpoint:])
    return _merge(left, right)


def _merge(left: list[int], right: list[int]) -> list[int]:
    result: list[int] = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] <= right[right_index]:
            result.append(left[left_index])
            left_index += 1
        else:
            result.append(right[right_index])
            right_index += 1

    result.extend(left[left_index:])
    result.extend(right[right_index:])
    return result


if __name__ == "__main__":
    sample = [42, 7, 19, 3, 7, 99, 1]
    print("MergeSort")
    print(f"input : {sample}")
    print(f"output: {merge_sort(sample)}")

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/sorting/MergeSort.java
package com.sortingsearching.sorting;

import java.util.Arrays;

public final class MergeSort {
    private MergeSort() {
    }

    public static int[] sort(int[] input) {
        int[] values = Arrays.copyOf(input, input.length);
        mergeSort(values, 0, values.length - 1);
        return values;
    }

    private static void mergeSort(int[] values, int left, int right) {
        if (left >= right) {
            return;
        }
        int mid = left + (right - left) / 2;
        mergeSort(values, left, mid);
        mergeSort(values, mid + 1, right);
        merge(values, left, mid, right);
    }

    private static void merge(int[] values, int left, int mid, int right) {
        int[] leftValues = Arrays.copyOfRange(values, left, mid + 1);
        int[] rightValues = Arrays.copyOfRange(values, mid + 1, right + 1);
        int i = 0;
        int j = 0;
        int k = left;

        while (i < leftValues.length && j < rightValues.length) {
            values[k++] = leftValues[i] <= rightValues[j] ? leftValues[i++] : rightValues[j++];
        }
        while (i < leftValues.length) {
            values[k++] = leftValues[i++];
        }
        while (j < rightValues.length) {
            values[k++] = rightValues[j++];
        }
    }

    public static void main(String[] args) {
        int[] sample = {42, 7, 19, 3, 7, 99, 1};
        System.out.println("MergeSort");
        System.out.println("input : " + Arrays.toString(sample));
        System.out.println("output: " + Arrays.toString(sort(sample)));
    }
}
